rskj-core/src/main/java/co/rsk/peg/whitelist/WhitelistStorageProviderImpl.java

```diff
@@ -0,0 +1,80 @@
+package co.rsk.peg.whitelist;
+
+import static co.rsk.peg.BridgeStorageIndexKey.LOCK_ONE_OFF_WHITELIST_KEY;
+import static co.rsk.peg.BridgeStorageIndexKey.LOCK_UNLIMITED_WHITELIST_KEY;
+import static org.ethereum.config.blockchain.upgrades.ConsensusRule.RSKIP87;
+
+import co.rsk.bitcoinj.core.Address;
+import co.rsk.bitcoinj.core.NetworkParameters;
+import co.rsk.peg.BridgeSerializationUtils;
+import co.rsk.peg.storage.StorageAccessor;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.commons.lang3.tuple.Pair;
+import org.ethereum.config.blockchain.upgrades.ActivationConfig;
+
+/**
+ * Implementation of {@link WhitelistStorageProvider} that saves the whitelist to the Storage and
+ * get the lock Whitelist.
+ */
+public class WhitelistStorageProviderImpl implements WhitelistStorageProvider {
+
+    private LockWhitelist lockWhitelist;
+    private final NetworkParameters networkParameters;
+    private final ActivationConfig.ForBlock activations;
+    private final StorageAccessor storageAccessor;
+
+    public WhitelistStorageProviderImpl(NetworkParameters networkParameters,
+        ActivationConfig.ForBlock activations,
+        StorageAccessor storageAccessor) {
+        this.networkParameters = networkParameters;
+        this.activations = activations;
+        this.storageAccessor = storageAccessor;
+    }
+
+    @Override
+    public void save() {
+        if (lockWhitelist == null) {
+            return;
+        }
+
+        List<OneOffWhiteListEntry> oneOffEntries = lockWhitelist.getAll(OneOffWhiteListEntry.class);
+        Pair<List<OneOffWhiteListEntry>, Integer> pairValue = Pair.of(oneOffEntries, lockWhitelist.getDisableBlockHeight());
+        storageAccessor.safeSaveToRepository(LOCK_ONE_OFF_WHITELIST_KEY.getKey(), pairValue, BridgeSerializationUtils::serializeOneOffLockWhitelist);
+
+        if (activations.isActive(RSKIP87)) {
+            List<UnlimitedWhiteListEntry> unlimitedEntries = lockWhitelist.getAll(UnlimitedWhiteListEntry.class);
+            storageAccessor.safeSaveToRepository(LOCK_UNLIMITED_WHITELIST_KEY.getKey(), unlimitedEntries, BridgeSerializationUtils::serializeUnlimitedLockWhitelist);
+        }
+    }
+
+    @Override
+    public LockWhitelist getLockWhitelist() {
```

Comment:
Good catch, @apancorb! Yes, it's a lazy initialization. WDYT if we use a synchronized block instead of synchronized on a method level?:

```
public LockWhitelist getLockWhitelist() {
        if (lockWhitelist == null) {
            synchronized (WhitelistStorageProviderImpl.class) {
                if (lockWhitelist == null) {
                    initializeLockWhitelist();
                }
            }
        }
        return lockWhitelist;
    }

    private void initializeLockWhitelist() {...}
```

Reply:
Hey, guys! @apancorb @marcos-iov 
I wanted to share with you some findings after applying the Double-Checked Locking(DCL). Despite that, DCL is widely cited and used as an efficient method for implementing lazy initialization in a multithreaded environment, after applying I faced a SonarLint warning that led me to get more information about DCL, and can help us make decisions in future similar situations:

> In early JVM versions, synchronizing entire methods was not performant, which sometimes caused this practice to use DCL in its place. However, given significant performance improvements of synchronized methods in recent JVM versions, synchronized methods are now preferred.

Apart from that, the recommendation is not to synchronize the private method but the `getLockWhitelist`. Something like that:

```
public synchronized LockWhitelist getLockWhitelist() {
        if (lockWhitelist == null) {
            lockWhitelist = initializeLockWhitelist();
        }
        return lockWhitelist;
    }

    private LockWhitelist initializeLockWhitelist() {
```

Reply:
Done! Please let me know if you want to discuss this topic deeper.

Reply:
Interesting @wilmerrootstock . Read operations don't need to be synchronized only the initialization aspect. Granular synchronized blocks are really not needed (they might even be more buggy than anything). But as you mentioned the performance penalty will not even have an effect in this case.